### PR TITLE
Feature/module arrange

### PR DESCRIPTION
**Pull request by lgannoaa**
Provide solution to issue #479 - Review modules being loaded on each platform (e.g. via module_base.*) to check they are loading in the proper order and dependencies are being met correctly. Check for module swaps and resolve as needed to maintain the correct module loads.
Change made to:
   OznMonBuild.orion, RadMonBuild.orion - Fix issue with resolving hpc-intel version. 
   module_base.orion - Fix issue with module swaps and arrange module for proper order.
   modulefile.ww3.orion - Fix issue with module swaps.
   module_base.wcoss_dell_p3 - Fix issue with missing version number in pm5 load. Fix issue with module swaps and arrange module for proper order.
   module_base.hera -  Fix issue with module swaps.
   modulefile.ww3.hera - Fix issue with module swaps.

### CODE REVIEW

**Comment by lgannoaa** [comment]
@KateFriedman-NOAA let me know if this draft PR look good to you. I will make it become a regular PR.

File: modulefiles/module_base.orion [hunk 1]
```diff
@@ -1,52 +1,41 @@
...
 module load ip/3.3.3
 module load nemsio/2.5.2
 module load sp/2.3.3
 module load w3emc/2.7.3
 module load w3nco/2.4.1
-module load upp/10.0.6
-
+module load jasper/2.0.25
+module load zlib/1.2.11
+module load png/1.6.35
```
Comment (aerorahul) [inline modulefiles/module_base.orion]: please move these back to their original location.  These are prerequisites to some other libraries that are loaded before.

File: modulefiles/module_base.orion [hunk 2]
```diff
@@ -1,52 +1,41 @@
...
+module load slurm/19.05.3-2
 
-module load jasper/2.0.25
-module load zlib/1.2.11
-module load png/1.6.35
+# waveprep
+module load cdo/1.9.5
 
-module load hdf5/1.10.6
-module load netcdf/4.7.4
```
Comment (aerorahul) [inline modulefiles/module_base.orion]: why are these removed?

File: modulefiles/module_base.orion [hunk 3]
```diff
@@ -1,52 +1,41 @@
...
-module load crtm/2.3.0
-setenv CRTM_FIX /apps/contrib/NCEPLIBS/orion/fix/crtm_v2.3.0
+module load hpc-impi/2018.4
+module load slurm/19.05.3-2
 
-module load jasper/2.0.25
-module load zlib/1.2.11
-module load png/1.6.35
+# waveprep
+module load cdo/1.9.5
```
Comment (aerorahul) [inline modulefiles/module_base.orion]: This module can only be loaded after hdf5 and netcdf are loaded.

Please do not move the module loads around.
The order of module loads is critical.

File: modulefiles/module_base.orion [hunk 4]
```diff
@@ -1,52 +1,41 @@
...
 module load esmf/8_1_1
 module load fms/2020.04.03
-
+module load grib_util/1.2.2
+module load prod_util/1.2.2
+module load gempak/7.5.1
 module load bacio/2.4.1
-module load g2/3.4.1
-module load g2tmpl/1.9.1
+module load upp/10.0.6
```
Comment (aerorahul) [inline modulefiles/module_base.orion]: UPP module requires several of the libraries loaded.
Please don't move the module loads around unless you have a sound reason to do so.

File: modulefiles/module_base.wcoss_dell_p3 [hunk 5]
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
...
 module load cdo/1.9.8
+
+# Workaround for github hpc-stack issues #370
+setenv LMOD_EXACT_MATCH no
+module load netcdf/4.7.4
+module load nemsio/2.5.2
+module load w3emc/2.7.3
+module load upp/10.0.6
+module load wgrib2/2.0.8
+setenv WGRIB2 wgrib2
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss_dell_p3]: Please move this back to where they were in the module load.

File: modulefiles/module_base.wcoss_dell_p3 [hunk 6]
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
 
 # waveprep
 module load cdo/1.9.8
+
+# Workaround for github hpc-stack issues #370
+setenv LMOD_EXACT_MATCH no
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss_dell_p3]: please set this variable immediately after `EnvVars` module is loaded.

File: modulefiles/module_base.wcoss_dell_p3 [hunk 7]
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
...
+module load netcdf/4.7.4
+module load nemsio/2.5.2
+module load w3emc/2.7.3
+module load upp/10.0.6
+module load wgrib2/2.0.8
+setenv WGRIB2 wgrib2
+setenv LMOD_EXACT_MATCH yes
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss_dell_p3]: This likely has no impact.

**Review by aerorahul — CHANGES_REQUESTED** [review]
Comments:
- versions updates will need to be tested. this is ok.
- the order in which modules are loaded is critical for ensuring a robust and reproducible software stack.  the order ensures prerequisites are available for subsequent modules.  please revert the order in which the modules are loaded, unless there was a bug encountered.

File: modulefiles/module_base.orion (hunk 3 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: Orion-login-1[1] /home/lgan$ module purge
Orion-login-1[2] /home/lgan$ module use /apps/contrib/NCEP/libs/hpc-stack/modulefiles/stack
Orion-login-1[3] /home/lgan$ module load hpc/1.1.0
Orion-login-1[4] /home/lgan$ module load hpc-intel/2018.4
Orion-login-1[5] /home/lgan$ module unload mkl/2020.2
Orion-login-1[6] /home/lgan$ module load mkl/2018.4
Orion-login-1[7] /home/lgan$
Orion-login-1[7] /home/lgan$ module load hpc-impi/2018.4
Orion-login-1[8] /home/lgan$ module load slurm/19.05.3-2
Orion-login-1[9] /home/lgan$ module load cdo/1.9.5
Orion-login-1[10] /home/lgan$ module list

Currently Loaded Modules:
  1) hpc/1.1.0   2) intel/2018.4   3) hpc-intel/2018.4   4) mkl/2018.4   5) impi/2018.4   6) hpc-impi/2018.4   7) munge/0.5.13   8) slurm/19.05.3-2   9) hdf5/1.10.6  10) netcdf/4.7.4  11) cdo/1.9.5

This shows that the hdf5 and netcdf is loaded after cdo module.

File: modulefiles/module_base.orion (hunk 4 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: load upp module also load g2 and g2tmpl automatically. That is the reason for the change.

File: modulefiles/module_base.orion (hunk 1 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: upp need to be load at the time g2 and g2tmpl is expected to be loaded. Because loading upp will also load g2 and g2tmpl automatically.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 6 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: LMOD_EXACT_MATCH=no effectively turn off all version check. It is not recommended practice.
The reason of setting LMOD_EXACT_MATCH=no is to provide a temporary workaround to address a known bug in wcoss_dell_p3 hpc-stack. Therefore, LMOD_EXACT_MATCH=no should only be used in the module impacted by this bug.

File: modulefiles/module_base.orion (hunk 1 above)
Comment (aerorahul) [reply modulefiles/module_base.orion]: This comment was for jasper, zlib and png.  How does upp figure in the response to this comment?

File: modulefiles/module_base.wcoss_dell_p3 (hunk 5 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: These module need not to be load with LMOD_EXACT_MATCH=no. Therefore they should be load before the LMOD_EXACT_MATCH=no statement.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 7 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: LMOD_EXACT_MATCH need to be turn on as default. That is the recommended best practice. Therefore, it is turned on after the temporary workaround solution has applied.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 6 above)
Comment (aerorahul) [reply modulefiles/module_base.wcoss_dell_p3]: `LMOD_EXACT_MATCH=no`, does not turn OFF version check.  It enables loading a `default` version designated, if a version is not specified.  It has no impact if a module is already loaded.

WCOSS2 does not require this any more.

File: modulefiles/module_base.orion (hunk 1 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: Answered in Files changed page.

File: modulefiles/module_base.orion (hunk 2 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: Answered in Files changed page.

File: modulefiles/module_base.orion (hunk 3 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: Answered in Files changed page.

File: modulefiles/module_base.orion (hunk 4 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: Answered in Files changed page.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 5 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: Answered in Files changed page.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 6 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3:50]: Answered in Files changed page.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 7 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: Answered in Files changed page.

File: modulefiles/module_base.wcoss_dell_p3 (hunk 6 above)
Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: @aerorahul 
This file module_base.wcoss_dell_p3 is not for wcoss2. 
Do you want me to set LMOD_EXACT_MATCH=no for the entire module_base.wcoss_dell_p3 file? 
Do you want me to not set LMOD_EXACT_MATCH=yes at the end of module_base.wcoss_dell_p3 file?

Comment (lgannoaa) [reply modulefiles/module_base.wcoss_dell_p3]: In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to set LMOD_EXACT_MATCH=no for the entire module_base.wcoss_dell_p3.

File: modulefiles/module_base.orion (hunk 3 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to module load netcdf and hdf5 before module load cdo. Keep module load cdo on the bottom of this file.

File: modulefiles/module_base.orion (hunk 4 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion]: In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to keep module load upp in the same location, keep module load g2 and g2tmpl in the same location, and update the version number for g2 and g2tmpl to match upp.

**Comment by lgannoaa** [comment]
@aerorahul, I updated the module file as you requested yesterday. Please take a look.

File: modulefiles/module_base.orion [hunk 8]
```diff
@@ -1,15 +1,16 @@
...
+# FV3GFS prerequisites
+########################################################
 
 module use /apps/contrib/NCEP/libs/hpc-stack/modulefiles/stack
 module load hpc/1.1.0
 module load hpc-intel/2018.4
+
+# gsi
+module unload mkl/2020.2
+module load mkl/2018.4
```
Comment (aerorahul) [inline modulefiles/module_base.orion:11]: why is this necessary here?  The GSI modulefile at build time does this switch.
Can you explain why/where this is used at runtime?

File: modulefiles/module_base.orion [hunk 9]
```diff
@@ -1,15 +1,16 @@
...
 module load hpc-impi/2018.4
 
-#module load hpss/hpss
```
Comment (aerorahul) [inline modulefiles/module_base.orion:11]: I can understand there is no `hpss` module.
Is `nco` not used on Orion?

File: modulefiles/module_base.wcoss_dell_p3 [hunk 10]
```diff
@@ -49,7 +52,6 @@ module load upp/10.0.6
 module load wgrib2/2.0.8
 setenv WGRIB2 wgrib2
 
-module use -a /gpfs/dell1/nco/ops/nwprod/modulefiles/
```
Comment (aerorahul) [inline modulefiles/module_base.wcoss_dell_p3:52]: are we sure we don't need this line?  How will `gempak` module be found? Is it in the default `MODULEPATH`?

**Review by aerorahul — CHANGES_REQUESTED** [review]
This PR is now reduced to the following items:

1. versions for a few libraries `g2`, `g2tmpl`, `w3emc` have been updated across `hera/orion/wcoss_dell_p3`.  Has this been robustly tested?  `wcoss_dell_p3` still uses a different `w3emc` library than on `Hera`?  What is the reason for this?
2. The Ozone and Radiance Monitoring build modulefiles should be deleted from this repository.  The GSI repository contains the relevant ones.  The ones in the global-workflow are obsolete.

There is no more "rearrangement" of the modules here as the PR title and description suggests.

File: modulefiles/module_base.orion (hunk 8 above)
Comment (lgannoaa) [reply modulefiles/module_base.orion:11]: Will contact GSI team for guidance.

Comment (RussTreadon-NOAA) [reply modulefiles/module_base.orion:11]: Background information is provided in NOAA-EMC/GSI issue [#110](https://github.com/NOAA-EMC/GSI/issues/110) and NOAA-EMC/hpc-stack issue [#170](https://github.com/NOAA-EMC/hpc-stack/issues/170).   @MichaelLueken-NOAA added to keep him in the loop.

**Comment by KateFriedman-NOAA** [comment]
This should probably be closed now. A number of the changed files in this draft PR have since been removed (e.g. OznMon/Radmon) or will be massively updated/renamed during the WCOSS2 port (LUA + hpc-stack updates). The module order is being addressed in the WCOSS2 port updates as well so this is a duplicate now. Thanks!
